### PR TITLE
OCE-47: Create a CI/CD pipeline to publish docs and artefacts

### PR DESCRIPTION
**Pull request by indigo423**
Upgrade CircleCI to use 2.1 executor features and streamline the setup. Introduce executors for:

* Building with a build environment
* Using the published Antora 2.0.0 image to compile the docs
* Use a netlify-cli container image to publish docs to Netlify
* Use a package-cloud-cli image to publish to netlify

Introduced a pre-build step which checks out the source code which allows us to set global build specific tasks and we can run steps easier in parallel. It removes the dependency in later executor images to have git and SSH client installed.

Changes required outside of this PR in CircleCI:

* Add an environment variable NETLIFY_AUTH_TOKEN to deploy docs
* Add an environment variable NETLIFY_SITE_ID which defines the target where docs gets deployed
* Add an environment variable PACKAGECLOUD_TOKEN to publish the RPMS to package cloud

I haven't changed the targets for Maven deployment to oss.sonatype and I haven't changed the targets for RPM deployment. The jobs `deploy-maven` and `deploy-packages` will fail, cause I don't have the correct credentials in my CircleCI project to publish to these repositories.

I've set a filter, build runs on every branch and deploy runs only triggered by `git push --tags`.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -1,44 +1,90 @@
...
+            tags:
+              only: /.*/
+      - build-docs:
+          requires:
+            - pre-build
+          filters:
+            tags:
+              only: /.*/
 
+      # Run deploy jobs just on version tags and ignore branches
```
Comment (j-white) [inline .circleci/config.yml]: We still want to deploy snapshot artifacts for the "master" branch (as we do currently)

File: test-api/.factorypath [hunk 2]
```diff
@@ -0,0 +1,6 @@
+<factorypath>
```
Comment (j-white) [inline test-api/.factorypath]: This needed?

**Comment by j-white** [comment]
Can we add this to a branch in the project and make sure it's working before we merge? Ideally I'd like to see documentation from `master` be published continuously as we work towards releasing 1.0.0.

File: test-api/.factorypath (hunk 2 above)
Comment (indigo423) [reply test-api/.factorypath]: This file is not from me, probably it was added after compiling and is a candidate from .gitignore. Should I add it there?

**Comment by indigo423** [comment]
>  Can we add this to a branch in the project and make sure it's working before we merge?
As I have no permissions, I can create a branch in the repository and can add Environment variables in Circle.

There are two things unknown to me:
* I've merged your Sonatype deployment, as soon I'm in the OCE repository, your configuration with credentials should be available. Is there anything wrong to deploy to Sonatype from a new created branch with the same version number 1.0.0-SNAPSHOT?
* Is there anything wrong deploying the same packages from the branch to Packagecloud which I've migrated from your master branch?

**Comment by j-white** [comment]
> * I've merged your Sonatype deployment, as soon I'm in the OCE repository, your configuration with credentials should be available. Is there anything wrong to deploy to Sonatype from a new created branch with the same version number 1.0.0-SNAPSHOT?

These artifacts aren't actively used, so no problems there.

> * Is there anything wrong deploying the same packages from the branch to Packagecloud which I've migrated from your master branch?

If the branch is out-of-date with master, it could cause problems, but it won't be up for long so there won't be any issues.

**Comment by indigo423** [comment]
Reopen PR from a branch from the repository itself.

**Pull request closed without merging**